src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PreferIsExactSpellingIsTrueForKnownApisAnalyzer.cs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Xml;
+using System.Xml.Serialization;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.NetCore.Analyzers.InteropServices
+{
+    internal static class PreferIsExactSpellingIsTrueForKownApisHelpers
+    {
+        internal static string GetActualExternName(this IMethodSymbol methodSymbol, AttributeData attribte)
+        {
+            var hasEntryPointParameter = attribte.NamedArguments.FirstOrDefault(x => x.Key.Equals("EntryPoint", StringComparison.Ordinal));
+            return hasEntryPointParameter.Key is null ? methodSymbol.Name : hasEntryPointParameter.Value.Value.ToString();
+        }
+    }
+
+    /// <summary>
+    /// CA1839: Prefer ExactSpelling=true for known Apis analyzer
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class PreferIsExactSpellingIsTrueForKnownApisAnalyzer : DiagnosticAnalyzer
+    {
+        [Serializable]
+#pragma warning disable CA1034 // Nested types should not be visible
+        public class KnownApi //Only made public as XmlSerializer requires it
+#pragma warning restore CA1034 // Nested types should not be visible
+        {
+            public string? Dll { get; set; }
+#pragma warning disable CA1819 // Properties should not return arrays
+            public string[]? Methods { get; set; } // DTO object may ignore this warning
+#pragma warning restore CA1819 // Properties should not return arrays
+        }
+
+        internal const string RuleId = "CA1839";
+
+        internal static readonly Lazy<Dictionary<string, HashSet<string>>> KnownApis = new Lazy<Dictionary<string, HashSet<string>>>(() => CreateKnownApis());
+
+        private static Dictionary<string, HashSet<string>> CreateKnownApis()
+        {
+            var serializer = new XmlSerializer(typeof(KnownApi[]));
+            var str = MicrosoftNetCoreAnalyzersResources.ResourceManager.GetObject("KnownApiList", CultureInfo.InvariantCulture);
+            using var stringReader = new StringReader(str.ToString());
+            using var reader = XmlReader.Create(stringReader);
+            var knownApis = (KnownApi[])serializer.Deserialize(reader);
```

Comment:
Why are we doing all these XML reader shenanigans? Can we please hard code the known APIs in source, just like all other analyzers do?

Reply:
Well there are currently 160k entries. I initially used the new generator feature to generate a compile time initialised dictionary with the constants, but the resulting compile time was an order of magnitude larger than compiling without the generated dictionary

Reply:
> Well there are currently 160k entries. 

We would absolutely need performance measurements with this analyzer enabled then. It seems the initial load be pretty expensive?

Reply:
Yes the initial load is the heaviest part, it is below half a second though and once its done it is very cheap to query

Reply:
Excluding name manged symbols this would shring to 96k method symbols